src/openlcb/IfTcpImpl.hxx

```diff
@@ -0,0 +1,553 @@
+        SequenceNumberGenerator *sequence)
+        : MessageStateFlowBase(service)
+        , sendTarget_(send_target)
+        , skipMember_(skip_member)
+        , gatewayId_(gateway_node_id)
+        , sequenceNumberGenerator_(sequence)
+    {
+    }
+
+private:
+    Action entry() override
+    {
+        auto id = nmsg()->dst.id;
+        if (id)
+        {
+            auto *n = iface()->lookup_local_node(id);
+            if (n)
+            {
+                // Addressed message loopback.
+                auto *b = transfer_message();
+                b->data()->dstNode = n;
+                iface()->dispatcher()->send(b, priority());
+                return release_and_exit();
+            }
+        }
+        return allocate_and_call(sendTarget_, STATE(render_src_message));
+    }
+
+    Action render_src_message()
+    {
+        auto b = get_buffer_deleter(get_allocation_result(sendTarget_));
+        TcpDefs::render_tcp_message(*nmsg(), gatewayId_,
+            sequenceNumberGenerator_->get_sequence_number(), b->data());
+        b->data()->skipMember_ = skipMember_;
+        sendTarget_->send(b.release(), nmsg()->priority());
+
+        // Checks and performs global loopback.
+        if (!nmsg()->dst.id)
+        {
+            iface()->dispatcher()->send(transfer_message(), priority());
+        }
+        return release_and_exit();
+    }
+
+    /// @return the abstract message we are trying to send.
+    GenMessage *nmsg()
+    {
+        return message()->data();
+    }
+
+    /// @return owning interface object.
+    If *iface()
+    {
+        return static_cast<If *>(service());
+    }
+
+    /// Where to send the rendered messages to.
+    HubPortInterface *sendTarget_;
+    /// This value will be populated to the skipMember_ field.
+    HubPortInterface *skipMember_;
+    /// Populated into the source gateway field of the outgoing messages.
+    NodeID gatewayId_;
+    /// Responsible for generating the sequence numbers of the outgoing
+    /// messages.
+    SequenceNumberGenerator *sequenceNumberGenerator_;
+};
+
+/// This flow is listening to data from a TCP connection, segments the incoming
+/// data into TcpMessages based on the incoing size, and forwards packets
+/// containing the TCP message as string payload.
+class FdToTcpParser : public StateFlowBase
+{
+public:
+    FdToTcpParser(FdHubPortService *s, HubPortInterface *dst,
+        HubPortInterface *skipMember)
+        : StateFlowBase(s)
+        , dst_(dst)
+        , skipMember_(skipMember)
+    {
+        HASSERT(s->fd() >= 0);
+        bufEnd_ = 0;
+        bufOfs_ = 0;
+        start_flow(STATE(start_msg));
+    }
+
+    /// Stops listening and terminates the flow.
+    void shutdown()
+    {
+        auto *e = this->service()->executor();
+        if (e->is_selected(&helper_))
+        {
+            e->unselect(&helper_);
+            helper_.remaining_ = 0;
+        }
+        set_terminated();
+        notify_barrier();
```

Comment:
This is an interesting pattern.  I wonder if it could be platformized and extended to support sleep_and_wait() states as well.  Just a thought.

Reply:
I've spent countless hours trying to make tests shut down cleanly. In the world of hubs and smarter gateways we have to bring down more and more complex objects in an appropriate manner. So framework support would be definitely nice.
We could use the otherwise currently empty Service concept to implement an ownership/guardian concept for a set of flows.
We can probably not make it happen automatically though; it is not possible to interrupt a state flow without giving it a chance to exit gracefully. The problem is that we never know if the flow is holding on to some resource or not or when there is a notification pointer in somebody else's hands that will come in later.